rfc/20240513-static-evaluation-providers.md

```diff
@@ -116,167 +117,325 @@ locals {
+      }))
+    }
+
+    resource "aws_vpc" "example" {
+      for_each = var.aws_regions
+      provider = each.value.provider_instance
+
+      cidr_block = each.value.cidr_block
+    }
+    ```
+
+    ```hcl
+    module "example" {
+      # ...
+
+      aws_regions = {
+        us-west-2 = {
+          cidr_block        = "10.1.0.0/16"
+          provider_instance = aws.usw2
+        }
+        eu-west-1 = {
+          cidr_block        = "10.2.0.0/16"
+          provider_instance = aws.euw2
+        }
+      }
+    }
+    ```
+
+##### Using the `count` meta-argument in `provider` blocks
+
+The existing `resource`, `data`, and `module` blocks support either `count` or `for_each` as two different strategies for causing a single configuration block to dynamically declare multiple instances.
+
+`count` is most appropriate for situations where the multiple instances are "fungible". For example, the instances could be considered "fungible" if the process of reducing the number of instances doesn't need to consider which of the instances will be destroyed. A collection of virtual machines all booted from the same disk image, with the same configuration, and running the same software could be considered fungible.
+
+`for_each` is the better option when each instance has at least one unique characteristic that makes it functionally or meaningfully distinct from the others. A collection of virtual network objects where each is used to host different services are _not_ fungible because each network has its own separate role to play in the overall system.
+
+The technical design we've adopted here could potentially allow supporting `count` in `provider` blocks, generating instances with integer numbers as keys just as with `count` in other blocks. However, we don't yet know of any clear use-case for a collection of "fungible" provider configurations: the only reasons we know of to have multiple instances of a provider involve each one being configured differently, such as using a different AWS region. It's important for correctness for the resource instances associated with these provider instances to retain their individual bindings as the set of provider instances changes, because an object created using one provider instance will often appear to be non-existent if requested with another provider instance.
+
+Because of how crucial it is to preserve the relationships between resource instances and provider instances between runs, we've chosen to intentionally exclude `count` support from the initial design. However, the `count` argument remains reserved in `provider` blocks so that we can potentially implement it in future if feedback suggests a significant use-case for it. We hope that the future discovered use-case(s) will also give us some ideas on how we could help protect against the inevitable misbehavior caused by a resource instance getting accidentally reassociated with the wrong provider instance.
```

Comment:
This was implemented in the implementation PR with a few lines.  It's easy to back out, but I'm not sure if it's more confusing to intentionally exclude it here.

Reply:
Given that everything we add in a stable release is forever, I would strongly recommend _not_ adding things that don't have a clear use-case just because they seem easy to implement.